DEP: Emit FutureWarning for NAT comparisons.

charris opened the pull request:
In Numpy 1.13 the plan is for NAT comparisons to behave like NaN
comparisons, e.g., False except for 'NAT != NAT', which will be
True. See the discussion at #7019 for details.

Needs tests but thought I would put this up for consideration. The question in my mind is whether to warn whenever a NAT appears, or only for NAT op NAT combinations where the value has changed. Currently it is the latter.

charris committed:
DEP: Emit FutureWarning for NAT comparisons.

In Numpy 1.13 the plan is for NAT comparisons to behave like NaN
comparisons, e.g., False except for 'NAT != NAT', which will be
True. See the discussion at gh-7019 for details.

charris commented:
It occurs to me that we should probably also warn in `numpy/testing/test_utils.py` as that may be one of the places most affected.

njsmith commented:
In general it's nice to make warning conditions as precise as possible, I
think, to make it clearer what's going on and reduce warning
fatigue/warning blindness?
On Jan 18, 2016 2:56 PM, "Charles Harris" <notifications@github.com> wrote:

> It occurs to me that we should probably also warn in
> numpy/testing/test_utils.py as that may be one of the places most
> affected.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7057#issuecomment-172675654>.
>

charris commented:
@njsmith So what is your recommendation here?

njsmith commented:
Sorry for being unclear :-) that was a vote for making the warning trigger
iff the user actually does a comparison whose result is affected.
On Jan 18, 2016 3:26 PM, "Charles Harris" <notifications@github.com> wrote:

> @njsmith <https://github.com/njsmith> So what is your recommendation here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7057#issuecomment-172680391>.
>

charris commented:
So that is what is currently here. The warnings issued here should show up in the tests also. I'll vote for that, less work ;)

charris committed:
TST: Add tests for NAT comparison FutureWarning.

The behavior of NAT comparisons will change in Numpy 1.13. Make sure
that a FutureWarning is emitted when the results will change.

shoyer commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -1126,11 +1126,49 @@ NPY_NO_EXPORT void
+            NPY_ALLOW_C_API_DEF
+            NPY_ALLOW_C_API;
+            /* 2016-01-18, 1.11 */
+            if (DEPRECATE_FUTUREWARNING(
```
Is there a reason for DEPRECATE_FUTUREWARNING rather than just DEPRECATE? Do we have a rule for when to choose which?

charris replied:
My understanding to the differences is this. A `DeprecationWarning` means that in the future either the feature will be removed or an exception will be raises. OTOH, a `FutureWarning` means that the feature will remain the its behavior will change.

I suppose the macro name `DEPRECATE_FUTUREWARNING` is a bit misleading, it just issues a `FutureWarning`.

shoyer commented:
@charris  thanks for putting this together! I would not have a figured out the GIL stuff on my own.

This looks good to me. I agree with issuing warnings only when necessary.

charris commented:
@shoyer I didn't know how to do it either, but I knew how to grep for it ;)

juliantaylor commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -1117,20 +1117,58 @@ NPY_NO_EXPORT void
+        const @type@ in2 = *(@type@ *)ip2;
+        *((npy_bool *)op1) = in1 != in2;
+
+        if (in1 == NPY_DATETIME_NAT && in1 == NPY_DATETIME_NAT) {
```
I assume this should be in2 in the second case?
this was not detected by our CI because apparently a change removed our compiler warning check ...

charris replied:
Yep, looks like a bug.

charris replied:
However, it still issues the required warning (which is why the test passed), it just issues it too often, even when the current comparison is already false.

charris replied:
@juliantaylor Could you open is issue or PR for this, milestone 1.12.1

juliantaylor replied:
see gh-8483
